lib/Sema/TypeCheckType.cpp

```diff
@@ -1325,6 +1260,31 @@ resolveTopLevelIdentTypeComponent(TypeResolution resolution,
     if (options.contains(TypeResolutionFlags::SilenceErrors))
       return ErrorType::get(ctx);
 
+    if (comp->getIdentifier() == ctx.Id_Self &&
+        !isa<GenericIdentTypeRepr>(comp)) {
+      DeclContext *nominalDC = nullptr;
+      NominalTypeDecl *nominal = nullptr;
+      auto dc = resolution.getDeclContext();
+      if ((nominalDC = dc->getInnermostTypeContext()) &&
+          (nominal = nominalDC->getSelfNominalTypeDecl())) {
+        assert(!isa<ProtocolDecl>(nominal) && "Cannot be a protocol");
+
+        Type SelfType = nominal->getDeclaredInterfaceType();
+
+// Causes assertion failure:
+// Assertion failed: (hasSelfMetadataParam() && "This method can only be called if the " "SILFunction has a self-metadata parameter"), function getSelfMetadataArgument, file /Volumes/Elements/swift-self/swift/include/swift/SIL/SILFunction.h, line 935.
```

Comment:
What test case triggers this error? It might be that SILGen already blows up trying to lower one of the invalid non-covariant examples we talked about.

Reply:
It was indeed `func foo(_ s: Self) -> Self` that was causing the crash. I've rejigged the code a little and Self outside the body of a class is only permitted for the return type now. We loose casting to Self inside classes in the process. Have a look at the changes to the tests in the last commit to see what the result is.

Reply:
Casting to Self inside classes is fine though. You should be able to differentiate type resolution inside a body from type resolution in a method signature by looking at the type resolution options.

Reply:
Thanks for the hint. I've added `|| options.isAnyExpr()` to the possibilities for Self on type resolution which fixes Casts in classes. It also allows `let a: Self = ...` to work. Is that ok?